### PR TITLE
Use temporary directory in test to avoid data race

### PR DESCRIPTION
**Pull request by mneverov**
Fixes: #7640 

Signed-off-by: Max Neverov <neverov.max@gmail.com>

The race happens because test uses the private manager targets field to [verify pool key presence](https://github.com/prometheus/prometheus/blob/master/discovery/manager_test.go#L910). All operations with this field are guarded [with the mutex](https://github.com/prometheus/prometheus/blob/master/discovery/manager.go#L150) and this invariant is violated in the test: there is no mutexes used.
But, there should not be any concurrent operations with the `targets`.
Because prometheus watches [not files, but a directory](https://github.com/prometheus/prometheus/blob/master/discovery/file/file.go#L205), and the test creates a config file directly in [/tmp](https://github.com/prometheus/prometheus/blob/master/discovery/manager_test.go#L862) it is possible to get events for the other test directories created in `/tmp` in other tests when run with `-race`.
By creating a directory devoted for this test the events from other tests should go away.

The other solution might be is to [watch over config files](https://github.com/prometheus/prometheus/blob/master/discovery/file/file.go#L215), not directories.

@roidelapluie could you please take a look?

### CODE REVIEW

File: discovery/manager_test.go [hunk 1]
```diff
@@ -859,11 +859,15 @@ func TestTargetSetRecreatesEmptyStaticConfigs(t *testing.T) {
 }
 
 func TestIdenticalConfigurationsAreCoalesced(t *testing.T) {
-	tmpFile, err := ioutil.TempFile("", "sd")
+	tmpDir, err := ioutil.TempDir("", "prometheus-cfg-coalesced")
+	if err != nil {
```
Comment (roidelapluie) [inline discovery/manager_test.go:863]:
```suggestion
	testutil.Ok(t, err)
```

File: discovery/manager_test.go [hunk 2]
```diff
@@ -859,11 +859,15 @@ func TestTargetSetRecreatesEmptyStaticConfigs(t *testing.T) {
 }
 
 func TestIdenticalConfigurationsAreCoalesced(t *testing.T) {
-	tmpFile, err := ioutil.TempFile("", "sd")
+	tmpDir, err := ioutil.TempDir("", "prometheus-cfg-coalesced")
+	testutil.Ok(t, err)
+		t.Fatalf("error creating temporary directory: %v", err)
```
Comment (roidelapluie) [inline discovery/manager_test.go:864]: then this needs to go and the next line too

Comment (mneverov) [reply discovery/manager_test.go]: thanks!
Replaced err check with the testutil in the test

**Comment by roidelapluie** [comment]
Can you rebase this? thanks

**Comment by mneverov** [comment]
@roidelapluie I think this is obsolete after #7701

**Comment by roidelapluie** [comment]
We have looked at this during the bug scrub and decided to close this as per previous comment. Thanks!

**Pull request closed without merging**